src/Algebra/Graph/LabelledAdjacencyMap/Internal.hs

```diff
@@ -0,0 +1,182 @@
+-----------------------------------------------------------------------------
+-- |
+-- Module     : Algebra.Graph.LabelledAdjacencyMap.Internal
+-- Copyright  : (c) Andrey Mokhov 2016-2018
+-- License    : MIT (see the file LICENSE)
+-- Maintainer : andrey.mokhov@gmail.com
+-- Stability  : unstable
+--
+-- This module exposes the implementation of labelled adjacency maps. The API is unstable
+-- and unsafe, and is exposed only for documentation. You should use the
+-- non-internal module "Algebra.Graph.LabelledAdjacencyMap" instead.
+-----------------------------------------------------------------------------
+module Algebra.Graph.LabelledAdjacencyMap.Internal
+        (
+    -- * Adjacency map implementation
+          LabelledAdjacencyMap(..)
+        , consistent
+        , edgeLabel
+        , empty
+        , connect
+        , overlay
+        , vertex
+        , (-<)
+        , (>-)
+        , fromAdjacencySets
+        )
+where
+
+import           Data.List
+import           Data.Map.Strict                ( Map
+                                                , keysSet
+                                                , fromSet
+                                                )
+import           Data.Set                       ( Set )
+import           Data.Maybe                     ( fromMaybe )
+import           Algebra.Graph.Label        ( Dioid(..)
+                                                , (\/)
+                                                , (/\)
+                                                , zero
+                                                )
+import           Control.DeepSeq                ( NFData(..) )
+import qualified Data.Map.Strict               as Map
+import qualified Data.Set                      as Set
+
+newtype LabelledAdjacencyMap a e = LAM {
+    -- | The /adjacency map/ of the graph: each vertex is associated with a set
+    -- of its direct successors.
+    labelledAdjacencyMap :: Map a (Map a e) } deriving Eq
+
+lAdjacencyMap :: LabelledAdjacencyMap a e -> Map a (Map a e)
+lAdjacencyMap = labelledAdjacencyMap
+
+
+
+instance (Ord a, Show a, Ord e, Show e, Dioid e) => Show (LabelledAdjacencyMap a e) where
```

Comment:
We need to figure out a way to visualise labelled graphs with edge labels.

Reply:
Yea, I'll try to come up with a way of doing this when I get a second.